### PR TITLE
fix: ads in article list for pagination and loading states

### PR DESCRIPTION
**Pull request by colincclark**
This fixes the lack of ads on second page and loading state ArticleLists. The ad slots need to be unique, and refreshed for it to work.

1. ArticleList is created in a loading state
2. Articles and Header load, and a new ad is shown
3. Clicking "next page" also renders a new ad

### CODE REVIEW

File: packages/article-list/src/article-list.web.js [hunk 1]
```diff
@@ -172,19 +175,28 @@ class ArticleList extends Component {
...
                 index > 0 ? <ArticleListItemSeparator /> : null;
 
-              const renderArticle = () => {
-                if (
-                  articlesLoading ||
-                  index !== 4 ||
-                  Object.keys(adConfig).length === 0
-                ) {
+              const renderAd = () => {
+                if (index !== 4 || Object.keys(adConfig).length === 0) {
```
[the remark this answers is not in the document]
Comment (colincclark) [reply packages/article-list/src/article-list.web.js]: done

File: packages/ad/src/dom-context.web.js [hunk 2]
```diff
@@ -25,6 +26,10 @@ export default class DOMContext extends React.PureComponent {
     this.processEventQueue();
   }
 
+  componentWillUnmount() {
+    this.processEvent = () => false;
```
Comment (craigbilner) [inline packages/ad/src/dom-context.web.js]: perhaps look at an alternative to overwriting the class method entirely. Flush the queue and guard on unmount?

File: packages/article-list/src/article-list.web.js [hunk 3]
```diff
@@ -172,19 +176,28 @@ class ArticleList extends Component {
...
+                  if (articlesLoading) {
+                    return "inline-articles-loading";
+                  }
+                  return "inline-ad";
+                };
+
+                const AdComponent = (
                   <AdComposer adConfig={adConfig}>
-                    <Ad pos="inline-ad" />
+                    <Ad counter={this.counter} pos={renderPosTag()} />
```
Comment (craigbilner) [inline packages/article-list/src/article-list.web.js]: could we change this to `slotPrefix` or something similar? As an API Ad shouldn't particularly care about it being a `counter` per se?

Comment (craigbilner) [inline packages/article-list/src/article-list.web.js]: `pos` could do with a more appropriate name

File: packages/article-list/src/article-list.web.js [hunk 4]
```diff
@@ -172,19 +176,28 @@ class ArticleList extends Component {
...
                 index > 0 ? <ArticleListItemSeparator /> : null;
 
-              const renderArticle = () => {
-                if (
-                  articlesLoading ||
-                  index !== 4 ||
-                  Object.keys(adConfig).length === 0
-                ) {
+              const renderAd = () => {
+                if (index !== this.advertPosition || Object.keys(adConfig).length === 0) {
```
Comment (craigbilner) [inline packages/article-list/src/article-list.web.js]: minor: potentially don't need to check the adConfig every render, just when it changes?

File: packages/article-list/src/article-list.web.js (hunk 3 above)
Comment (aronshamash) [reply packages/article-list/src/article-list.web.js]: It looks like that is a name the ad guys have used for targeting so the prop name is just following that.

Comment (colincclark) [reply packages/article-list/src/article-list.web.js]: done

Comment (colincclark) [reply packages/article-list/src/article-list.web.js]: `slot.setTargeting("pos", containerID);` - but I can still change the JS variables

Comment (colincclark) [reply packages/article-list/src/article-list.web.js]: done

File: packages/article-list/src/article-list.web.js (hunk 4 above)
Comment (colincclark) [reply packages/article-list/src/article-list.web.js]: moved outside of the loop

**Pull request merged**